src/combinatorics/hales_jewett.lean

```diff
@@ -0,0 +1,299 @@
+be nonempty in the statement of `exists_mono_in_high_dimension`.
+
+We do not prove the finitary version of Van der Waerden's theorem (which is equivalent to the
+infinite one by compactness), although it follows from the same proof.
+
+In principle our proof should give computable upper bounds on the numbers `HJ(m, k)`, but this would
+require rewriting the proof to avoid `Prop`, and the upper bounds are enormous anyway.
+
+## Main results
+
+- `combinatorics.line.exists_mono_in_high_dimension m k`: the Hales-Jewett theorem.
+- `combinatorics.exists_mono_homothetic_copy M S f k C`: a generalization of Van der Waerden's
+theorem to an arbitrary commutative monoid. The proof is a simple application of Hales-Jewett.
+
+## Tags
+
+combinatorial line, Ramsey theory, arithmetic progession
+
+### References
+
+* https://en.wikipedia.org/wiki/Hales%E2%80%93Jewett_theorem
+
+Note that this file does not use the same notation as Wikipedia.
+
+-/
+
+open_locale classical
+open_locale big_operators
+
+instance (n : pnat) : nonempty (fin n) := ⟨⟨0, n.pos⟩⟩
+
+/-- An induction principle for types analogous to `pnat.rec_on`. To prove a statement about
+all nonempty finite types, it suffices to consider `unit` and `option α` where we know the
+statement for `α`, provided that our statement is preserved by `≃`.-/
+lemma fintype_induction {P : Type → Prop} (of_equiv : ∀ {α β}, α ≃ β → P α → P β)
+  (h_unit : P unit) (h_option : ∀ {α} [fintype α] [nonempty α], P α → P (option α))
+  (α) [fintype α] [nonempty α] : P α :=
+begin
+  suffices : ∀ n : pnat, P (fin n),
+  { refine of_equiv (fintype.equiv_fin α).symm
+      (this ⟨fintype.card α, finset.card_pos.mpr finset.univ_nonempty⟩), },
+  intro n, apply n.rec_on,
+  exact of_equiv fin_one_equiv.symm h_unit,
+  exact λ _ ih, of_equiv (fin_succ_equiv _).symm (h_option ih),
+end
+
+namespace combinatorics
+
+/-- The type of combinatorial lines. A line `l : line m n` in the hypercube `n → m` defines a
+function `m → (n → m)` to the hypercube, such that for each coordinate `i : n`, the function
+`l x i` is either identically `x` or constant. We require lines to be nontrivial, in the sense
+that `l x i` is identically `x` for at least one `i`.
+
+Formally, a line is represented by the function `l.idx_fun : n → option m` which says whether
+`l x i` is identically `x` (corresponding to `l.idx_fun i = none`) or constantly `y`
+(corresponding to `l.idx_fun i = some y`). When `m` has size `1` there can be many elements of
+`line m n` defining the same function. -/
+structure line (m n : Type) : Type :=
+(idx_fun : n → option m)
+(proper : ∃ i, idx_fun i = none)
+
+namespace line
+
+/- This lets us treat a line `l : line m n` as a function `m → (n → m)`. -/
+instance (m n) : has_coe_to_fun (line m n) :=
+⟨λ _, m → n → m, λ l x i, (l.idx_fun i).get_or_else x⟩
+
+/-- The type of lines that are only one color except possibly at their endpoints. -/
+@[nolint has_inhabited_instance]
```

Comment:
I won't insist upon it, but it would be nice to give an example here, even if it's a trivial one

Reply:
I think this is tricky because there will only be an example for certain colourings. You could say something tautological like "if there is a line then it is almost mono wrt a constant coloring", but I don't know if this is helpful.

Reply:
This is fair - but that's usually fine for the inhabited linter